electron/src/js/lib/SingleSignOn.ts

```diff
@@ -0,0 +1,318 @@
+        if (error) {
+          throw new Error(`Failed to register protocol. Error: ${error}`);
+        }
+      });
+    },
+    unregister: (session: Electron.Session): Promise<void> => {
+      return new Promise((resolve, reject) => {
+        session.protocol.unregisterProtocol(SingleSignOn.SSO_PROTOCOL, error => {
+          if (error) {
+            reject(error);
+          }
+          resolve();
+        });
+      });
+    },
+  };
+
+  // Ensure authenticity of the window from within the code
+  public static isSingleSignOnLoginWindow = (frameName: string) => SingleSignOn.SINGLE_SIGN_ON_FRAME_NAME === frameName;
+
+  // Ensure the requested URL is going to the backend
+  public static isBackendOrigin = (url: string) => SingleSignOn.ALLOWED_BACKEND_ORIGINS.includes(new URL(url).origin);
+
+  public static readonly javascriptHelper = () => {
+    return `Object.defineProperty(window, 'opener', {
+      configurable: true, // Needed on Chrome :(
+      enumerable: false,
+      value: Object.freeze({
+        postMessage: ({type}) => {
+          const params = new URLSearchParams();
+          params.set('secret', '${SingleSignOn.loginAuthorizationSecret}');
+          params.set('type', type);
+          const url = new URL('${SingleSignOn.SSO_PROTOCOL}://${SingleSignOn.SSO_PROTOCOL_HOST}/');
+          url.search = params.toString();
+          document.location.href = url.toString();
+        }
+      }),
+      writable: false,
+    });`;
+  };
+
+  constructor(
+    private readonly mainBrowserWindow: Electron.BrowserWindow,
+    private readonly senderEvent: Electron.Event,
+    windowOriginUrl: string,
+    private readonly windowOptions: Electron.BrowserWindowConstructorOptions
+  ) {
+    this.mainSession = this.mainBrowserWindow.webContents.session;
+    this.senderWebContents = senderEvent.sender;
+    this.windowOriginUrl = new URL(windowOriginUrl);
+  }
+
+  public readonly init = async (): Promise<void> => {
+    // Create a ephemeral and isolated session
+    this.session = session.fromPartition(SingleSignOn.SSO_SESSION_NAME, {cache: false});
+
+    // Disable browser permissions (microphone, camera...)
+    this.session.setPermissionRequestHandler((webContents, permission, callback) => callback(false));
+
+    const SingleSignOnLoginWindow = this.createBrowserWindow();
+
+    // Register protocol
+    // Note: we need to create the window before otherwise it does not work
+    await SingleSignOn.protocol.register(this.session, (responseType: string) => this.finalizeLogin(responseType));
+
+    // Show the window(s)
+    SingleSignOnLoginWindow.loadURL(this.windowOriginUrl.toString());
+    if (argv.devtools) {
+      SingleSignOnLoginWindow.webContents.openDevTools({mode: 'detach'});
+    }
+  };
+
+  private readonly createBrowserWindow = (): Electron.BrowserWindow => {
+    // Discard old preload URL
+    delete (<any>this.windowOptions).webPreferences.preloadURL;
```

Comment:
Can we get rid of `any`?

Reply:
Unfortunately not, preload is present in the typing definitions but preloadURL is not.

Reply:
But then `webPreferences` should be of type `any`, right?

Reply:
For some reason it's not:

```
electron/src/js/lib/SingleSignOn.ts:235:46 - error TS2339: Property 'preloadURL' does not exist on type 'WebPreferences'.

235     delete this.windowOptions.webPreferences.preloadURL;
                                                 ~~~~~~~~~~
```